JOSE_paper/paper.md

```diff
@@ -0,0 +1,223 @@
+
+- Introduction to research software engineering, why it matters, and some examples,
+- The use of virtual environments and specifying project dependencies,
+- Code formatting for consistent readable layout,
+- Naming principles for readable, re-useable code,
+- Linting and static analysis of code,
+- Writing docstrings and comments, and
+- Other general principles including magic numbers, avoiding hardcoded values, etc.
+
+This makes it easy to remove sections to tailor the course, or for additional topics to be added in future.
+
+## Exercises
+
+The slides are interspersed with a series of exercises, each one applying techniques
+that have just been taught.
+
+We set up a scenario where users have been provided a working, but poorly written piece
+of code that they are now expected to re-use in their work.
+As they progress through the exercises they apply their new skills to transform this
+into a much more readable and understandable piece of code.
+
+The script we use is based on some of the code and data available in @Irving2019.
+This was chosen as it features typical applications we expect users to encounter
+(reading from a dataset, processing, and plotting).
+The course was originally designed to be taught to geoscientists, and then at a climate
+science summer school, and we wanted participants to be focussed on the changes to the
+code rather than worrying about 
+
+Exercises roughly match the slide sections above:
+
+- 01: Baseline code to examine, install dependencies, and run.\
+  Users can familiarise themselves with what the code does, see why listing dependencies
+  with a quick setup is useful, and make early observations about how the code is
+  difficult to understand/use.
+- 02: Apply a formatter to standardise code.\
+  We use the black python formatter [@LangaBlack]
+- 03: Improve code clarity with naming and source changes.
+- 04: Linting and static analysis of code.
+- 05: Writing docstrings and best use of comments.
+- 06: General techniques for better code (magic numbers, string formatting).
+- 00: The end point of the workshop - an improved version of the code in exercise 01.
+
+
+### Solutions
+
+Worked solutions to all of the exercises are provided in the form of completed
+notebooks including example output.
+These are available both in the repository and also as Colab instances.
+
+
+
+# Content Delivery
+
+The course has been designed to be flexible in terms of delivery, allowing
+it to be adapted to and reused in various setups.
+
+TODO Slides are interspersed with exercises.
+
+The main aspect we wish to emphasise in delivery is teaching in a _"code-along"_ fashion.
+This helps with engagement, participation, and understanding [@barba2022teaching] and
+is essential, we feel, to having a long-lasting benefit.
+This approach slows those leading the course towards the rate at which the participants
+are working, and illustrates through errors (whether intentional or not!) that even
```

Comment:
Could cite https://dl.acm.org/doi/10.1145/2445196.2445388 about the effectiveness of live coding

Reply:
Done, see https://github.com/jatkinson1000/power-up-python/pull/14/commits/2816d304cbae61641b9a2d4d75ea7fc8b2954b4d